server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesEncoder.java

```diff
@@ -187,23 +187,57 @@ void encode(long[] in, DataOutput out) throws IOException {
      *     <li>byte 0: 1/2 bits header+6/7 bits data</li>
      *     <li>byte 1..n: data</li>
      * </ul>
-     * The header (first 1 or 2 bits) describes how the data is encoded:
+     * The header (first 1-3 bits) describes how the data is encoded:
      * <ul>
-     *     <li>?0 block has a single value (vlong), 2nd bit already contains data</li>
+     *     <li><code>?0</code>: block has a single value (vlong), 2nd bit already contains data</li>
      *     <li>
-     *         01 block has two runs, data contains value 1 (vlong), run-length (vint) of value 1,
+     *         <code>01</code>: block has two runs, data contains value 1 (vlong), run-length (vint) of value 1,
      *         and delta from first to second value (zlong)
      *     </li>
-     *     <li>11 block is bit-packed</li>
+     *     <li>
+     *         <code>111</code>: block contains cyclic data, data contains cycle length (vlong),
+     *         and the values until the cycle repeats (encoded as vlongs)
+     *     </li>
+     *     <li><code>011</code>: block is bit-packed</li>
      * </ul>
      */
     void encodeOrdinals(long[] in, DataOutput out, int bitsPerOrd) throws IOException {
         assert in.length == ES87TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE;
         int numRuns = 1;
+        long firstValue = in[0];
+        long previousValue = firstValue;
+        boolean cyclic = false;
+        int cycleLength = 0;
         for (int i = 1; i < in.length; ++i) {
-            if (in[i - 1] != in[i]) {
+            long currentValue = in[i];
+            if (previousValue != currentValue) {
                 numRuns++;
             }
+            if (currentValue == firstValue && cycleLength != -1) {
+                if (cycleLength == 0) {
+                    // first candidate cycle detected
+                    cycleLength = i;
+                } else if (i % cycleLength != 0) {
+                    // this isn't a cycle if the index of the next occurrence of the first value
+                    // isn't a multiple of the candidate cycle length
+                    // we can stop looking for cycles
+                    cycleLength = -1;
+                }
```

Comment:
For simplicity, maybe we could skip this check and rely on the code a few lines below that checks whether there is actually a cycle.

Reply:
Fine with me.